### ISSUE TITLE
boost_from_average does not work with custom metric

### ISSUE THREAD

**Issue by rydevera3** [microsoft/LightGBM#2558]
Hello all,

I created a custom metric but `boost_from_average` does not work if I set it to `True`. Is there a trick to getting it to work with a custom metric? Here is a reproducible example:

```
import numpy as np
from sklearn import datasets
from sklearn.metrics import confusion_matrix, log_loss
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
import lightgbm as lgb

# Iris dataset
iris = datasets.load_iris()
X, Ymc = iris.data, iris.target
X_train, X_test, y_train, y_test = train_test_split(X, Ymc, test_size=0.2,
                                                    random_state=2019)

lgb_train = lgb.Dataset(X_train, y_train)
lgb_test = lgb.Dataset(X_test, y_test)

def logregobj_lgb(preds, train_data, label_smoothing=0.0):
    
    labels = train_data.get_label()

    preds = np.reshape(preds, (len(labels), 3), 'F')
    preds = np.exp(preds - np.max(preds))
    preds = np.multiply(preds, 1 / np.sum(preds, axis=1)[:, np.newaxis])
    
    labels = OneHotEncoder(categories='auto', sparse=False).fit_transform(labels.reshape(-1, 1))
    
    grad = (preds - labels)
    hess = 2.0 * preds * (1.0 - preds)
    return grad.flatten('F'), hess.flatten('F')

def multi_logloss(preds, train_data):
    
    labels = train_data.get_label()
    preds = np.reshape(preds, (len(labels), 3), 'F')
    preds = np.exp(preds - np.max(preds))
    preds = np.multiply(preds, 1 / np.sum(preds, axis=1)[:, np.newaxis])
    
    np.argmax(preds, axis=1)
    print(preds)
    
    return 'multi_logloss', log_loss(labels, preds), False

def test_metric(preds, train_data):
    
    labels = train_data.get_label()
    preds = np.reshape(preds, (len(labels), 3), 'F')
    print(preds)
    
    return 'multi_logloss', log_loss(labels, preds), False
```

Then I implement without the custom metric to observe the predictions:

```
#train lightgbm with default multiclass loss function
params={'num_class':len(np.unique(Ymc)), 'objective': 'multiclass',
        'num_leaves': 63, 'num_iterations': 1, 'boost_from_average': True}

bst=lgb.train(params, lgb_train, verbose_eval=1,
              valid_sets=lgb_test,
              feval=test_metric)

[[0.36931694 0.33011871 0.30056435]
 [0.36931694 0.33011871 0.30056435]
 [0.29301589 0.33199833 0.37498578]
 [0.29195585 0.4086758  0.29936835]
 [0.29301589 0.33199833 0.37498578]
 [0.36931694 0.33011871 0.30056435]
 [0.29301589 0.33199833 0.37498578]
 [0.36931694 0.33011871 0.30056435]
 [0.29296091 0.39262987 0.31440922]
 [0.29301589 0.33199833 0.37498578]
 [0.29296091 0.39262987 0.31440922]
 [0.29301589 0.33199833 0.37498578]
 [0.29301589 0.33199833 0.37498578]
 [0.29195585 0.4086758  0.29936835]
 [0.36931694 0.33011871 0.30056435]
 [0.29195585 0.4086758  0.29936835]
 [0.29301589 0.33199833 0.37498578]
 [0.29296091 0.39262987 0.31440922]
 [0.36931694 0.33011871 0.30056435]
 [0.29301589 0.33199833 0.37498578]
 [0.36931694 0.33011871 0.30056435]
 [0.29301589 0.33199833 0.37498578]
 [0.36931694 0.33011871 0.30056435]
 [0.29296091 0.39262987 0.31440922]
 [0.36931694 0.33011871 0.30056435]
 [0.36931694 0.33011871 0.30056435]
 [0.29296091 0.39262987 0.31440922]
 [0.29301589 0.33199833 0.37498578]
 [0.36931694 0.33011871 0.30056435]
 [0.36931694 0.33011871 0.30056435]]
```

Then I use my custom metric:

```
bst=lgb.train(params, lgb_train, verbose_eval=1, 
              fobj=logregobj_lgb, valid_sets=lgb_test,
              feval=multi_logloss)

[[0.38505488 0.30747256 0.30747256]
 [0.38505488 0.30747256 0.30747256]
 [0.30761756 0.30954619 0.38283624]
 [0.30747256 0.38505488 0.30747256]
 [0.30761756 0.30954619 0.38283624]
 [0.38505488 0.30747256 0.30747256]
 [0.30761756 0.30954619 0.38283624]
 [0.38505488 0.30747256 0.30747256]
 [0.30833314 0.3691418  0.32252506]
 [0.30761756 0.30954619 0.38283624]
 [0.30833314 0.3691418  0.32252506]
 [0.30761756 0.30954619 0.38283624]
 [0.30761756 0.30954619 0.38283624]
 [0.30747256 0.38505488 0.30747256]
 [0.38505488 0.30747256 0.30747256]
 [0.30747256 0.38505488 0.30747256]
 [0.30761756 0.30954619 0.38283624]
 [0.30833314 0.3691418  0.32252506]
 [0.38505488 0.30747256 0.30747256]
 [0.30761756 0.30954619 0.38283624]
 [0.38505488 0.30747256 0.30747256]
 [0.30761756 0.30954619 0.38283624]
 [0.38505488 0.30747256 0.30747256]
 [0.30833314 0.3691418  0.32252506]
 [0.38505488 0.30747256 0.30747256]
 [0.38505488 0.30747256 0.30747256]
 [0.30833314 0.3691418  0.32252506]
 [0.30761756 0.30954619 0.38283624]
 [0.38505488 0.30747256 0.30747256]
 [0.38505488 0.30747256 0.30747256]]
```

The outputs are equal when `boost_from_average=False`.

**Comment by guolinke** [comment]
@rydevera3 is there any warning information?

**Comment by guolinke** [comment]
@rydevera3 okay, the root cause the customed objective, not the customed metric.
As boost_from_average has dependence over objective function, so it is not supported for customed objective. You can use `init_score` to set your own initial score, based on your objective functions.

**Comment by rydevera3** [comment]
Thank you @guolinke for the info. Do you have an example of how to use `init_score` for the multi-class classification case? Ideally, I'd just like to be able to replicate the boost from average functionality and then I can extend the objective function in the way I need to for my project.

**Comment by guolinke** [comment]
refer to https://github.com/microsoft/LightGBM/blob/master/src/objective/multiclass_objective.hpp#L150

class_init_probs_ is the distribution of classes over training data.
